### PR TITLE
Speculative execution fixes

### PR DESCRIPTION
**Pull request by Dieterbe**

### CODE REVIEW

File: api/cluster.go [hunk 1]
```diff
@@ -382,12 +388,12 @@ func (s *Server) peerQuerySpeculative(ctx context.Context, data cluster.Traceabl
 			delete(pendingResponses, resp.shardGroup)
 			delete(originalPeers, resp.data.peer.GetName())
 
-		case <-specCheckTicker.C:
+		case <-tickChan:
 			// Check if it's time to speculate!
-			percentReceived := 1 - (float64(len(pendingResponses)) / float64(len(peerGroups)))
-			if percentReceived > speculationThreshold {
+			percentReceived := float64(len(receivedResponses)) / float64(len(peerGroups))
```
Comment (shanson7) [inline api/cluster.go]: Is `pendingResponses` still worth anything? It seems superfluous. The current uses are:
1. To drive the main loop (`for len(pendingResponses) > 0`) which could be `for len(receivedResponses) < len(peerGroups)`
2. To decide upon who to speculate which could loop on `peerGroups` and continue if in `recievedResponses`. 

I'm not sure it's worth the extra allocation.

File: api/cluster.go [hunk 2]
```diff
@@ -363,7 +363,13 @@ func (s *Server) peerQuerySpeculative(ctx context.Context, data cluster.Traceabl
 
 	result := make(map[string]PeerResponse)
 
-	specCheckTicker := time.NewTicker(5 * time.Millisecond)
+	var ticker *time.Ticker
+	var tickChan <-chan time.Time
+	if speculationThreshold != 1 {
```
Comment (replay) [inline api/cluster.go]: that's nice. so the ticker doesn't even get instantiated if `speculationThreshold == 1`

File: api/cluster.go (hunk 1 above)
Comment (replay) [inline api/cluster.go]: that's changing the logic, but it's actually making it correct, so that's good.

Comment (shanson7) [reply api/cluster.go]: What's the change in logic?

Comment (Dieterbe) [reply api/cluster.go]: this particular line should not change logic. but the line below it does.
i explained in depth in the commit message
(i generally try to make commit messages explain everything)

Comment (shanson7) [reply api/cluster.go]: Yes, that was my analysis as well 👍

**Pull request merged**